algebra2/gaussian_elimination/gaussian_elimination.h

```diff
@@ -0,0 +1,566 @@
+    auto show_steps(ranges::matrix_view<T, LP> m,
+                    axis a,
+                    operation allowed_operations = operation::swap | operation::add | operation::multiply,
+                    reducted_form reducted = reducted_form::echelon) -> void;
+
+    enum class error {
+        not_invertible,
+        not_square,
+    };
+
+    // returns inverse of matrix m
+    template <typename T, typename LP>
+    auto inverse(ranges::matrix_view<T, LP> m) -> std::expected<ranges::matrix_view<T, LP>, algorithms::gaussian_elimination::error>;
+
+    /*calculates the determinant of matrix m*/
+    template <typename T, typename LP>
+    auto determinant(ranges::matrix_view<const T, LP> m) -> std::expected<T, error>;
+
+    /*calculates the determinant of matrix m containing fractions*/
+    template <std::integral T, typename LP>
+    auto determinant(ranges::matrix_view<ranges::fraction<T>, LP> m) -> std::expected<ranges::fraction<T>, algorithms::gaussian_elimination::error>;
+
+}  // namespace algorithms::gaussian_elimination
+
+
+/*description:
+ *      Performs gaussian elimination on matrix m containing fractions*/
+template<std::integral T, typename LP>
+auto run(ranges::matrix_view<ranges::fraction<T>, LP> m,
+         algorithms::gaussian_elimination::axis a,
+         algorithms::gaussian_elimination::operation allowed_operations = algorithms::gaussian_elimination::operation::swap | algorithms::gaussian_elimination::operation::add |
+                 algorithms::gaussian_elimination::operation::multiply,
+         algorithms::gaussian_elimination::reducted_form reducted = algorithms::gaussian_elimination::reducted_form::echelon) -> void {
+
+    std::println("//////////Gaussian Elimination on matrix m containing fractions//////////\n");
+    std::println("Matrix m = {}\n", m);
+    ranges::fraction<T> res;
+
+    for(int i = 0;  i < m.extent(0); i++){
+        if(m[i][i].numerator == 0){
+            for(int j = i+1; j < m.extent(0); j++){
+                if(m[j][i].numerator != 0){
+                    swap(m, i, j);
+                }
+            }
+        }
+        if(m[i][i].numerator != 0){
+            for(int k = i+1; k < m.extent(0); k++){
+                res = m[k][i]/m[i][i];
+                if(res.numerator != 0){
+                    subtract(m, k, i, res);
+                }
+            }
+        }
+    }
+
+    if(reducted == algorithms::gaussian_elimination::reducted_form::diagonal){
+        for(int i = 0;  i < m.extent(0); i++){
+            if(m[i, i].numerator != 0){
+                for(int k = i+1; k < m.extent(0); k++){
+                    res = m[i, k]/m[k, k];
+                    if(res.numerator != 0){
+                        subtract(m, i, k, res);
+                    }
+                }
+            }
+        }
+    }
+
+    std::println("The reduced matrix is {}\n", m);
+}
+
+template<typename T, typename LP>
+auto run(ranges::matrix_view<T, LP> m,
```

Comment:
repetition of code - use previous implementation of `run` function to handle this overload